crates/compiler/mono/src/layout.rs

```diff
@@ -4134,9 +4134,16 @@ where
         Unit => env
             .cache
             .put_in(Layout::new(LayoutRepr::UNIT, compute_semantic())),
-        BoolUnion { .. } => env
-            .cache
-            .put_in(Layout::new(LayoutRepr::BOOL, compute_semantic())),
+        BoolUnion { .. } => {
+            let semantic = compute_semantic();
+
+            // deduplicate the `Bool : [ True, False ]` type representation
+            if semantic == SemanticRepr::BOOL {
+                Layout::BOOL_SEMA
+            } else {
+                env.cache.put_in(Layout::new(LayoutRepr::BOOL, semantic))
+            }
+        }
```

Comment:
If we want to short-circuit on deduplicating the `Bool` type, it may be better to not call `compute_semantic`, and just check the tags directly, since it avoids an allocation. I think that's the biggest benefit of this PR, right?